torch/distributed/elastic/utils/gpu_health_check.py

```diff
@@ -0,0 +1,382 @@
+            on_failure (Optional[Callable]): Callback function to handle health check failures.
+        """
+        super().__init__()
+        self.device_index = device_index
+        self.interval = interval
+        self.on_failure = on_failure
+        self.pynvml_available = self.check_pynvml_availability()
+        self.enabled = self._check_driver_version()
+
+    @with_pynvml_lock
+    def _check_driver_version(self) -> bool:
+        """
+        Checks if the GPU driver version supports health checks (version r570 or newer).
+
+        Returns:
+            bool: True if the driver supports health checks, False otherwise.
+        """
+        GPU_RECOVERY_API_MIN_DRIVER_VERSION = 570
+
+        if not self.pynvml_available:
+            logger.warning("GPU Health checks are disabled because pynvml is not available.")
+            return False
+
+        try:
+            self.pynvml.nvmlInit()
+            driver_version = self.pynvml.nvmlSystemGetDriverVersion()
+            if isinstance(driver_version, bytes):
+                driver_version = driver_version.decode('utf-8')
+            self.pynvml.nvmlShutdown()
+
+            major_version = int(driver_version.split('.')[0])
+
+            if major_version < GPU_RECOVERY_API_MIN_DRIVER_VERSION:
+                logger.warning(
+                    f"Health checks disabled: GPU driver version r{major_version} is older than "
+                    f"required r{GPU_RECOVERY_API_MIN_DRIVER_VERSION} for the GPU Recovery API."
+                )
+                return False
+            return True
+
+        except Exception as e:
+            logger.warning(
+                f"GPU Health checks disabled: Unable to determine driver version due to: {e}"
+            )
+            return False
+
+    async def async_check(self) -> None:
+        """
+        Asynchronous GPU health check that runs periodically.
+
+        Periodically checks GPU health and handles any failures if they occur.
+        """
+        if not self.enabled:
+            return
+
+        while True:
+            await asyncio.sleep(self.interval)
+            result = await self._check_health()
+            if not result and self.on_failure:
+                await self.on_failure()
+
+    async def _check_health(self) -> bool:
+        """
+        Performs the asynchronous GPU health check.
+
+        Returns:
+            bool: True if all GPUs are healthy, False if any GPU has an issue.
+        """
+        return self._perform_health_check()
+
+    def __call__(self) -> Union[Optional[Exception], bool]:
+        """
+        Synchronous GPU health check callable.
+
+        Returns:
+            bool: Returns True if GPUs are healthy.
+        """
+        if not self.enabled:
+            logger.warning("Health checks are disabled; skipping synchronous check.")
+            return True
+
+        result = self._perform_health_check()
+        return result
```

Comment:
Thoughts on making this always return some form of Exception? Bool output means we can't report failure info correctly using normal torchelastic logging mechanisms